contracts/l2/discovery/L2GNS.sol

```diff
@@ -0,0 +1,315 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import { SafeMathUpgradeable } from "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+
+import { AddressAliasHelper } from "../../arbitrum/AddressAliasHelper.sol";
+import { GNS } from "../../discovery/GNS.sol";
+import { IGNS } from "../../discovery/IGNS.sol";
+import { ICuration } from "../../curation/ICuration.sol";
+import { IL2GNS } from "./IL2GNS.sol";
+import { L2GNSV1Storage } from "./L2GNSStorage.sol";
+
+import { IL2Curation } from "../curation/IL2Curation.sol";
+
+/**
+ * @title L2GNS
+ * @dev The Graph Name System contract provides a decentralized naming system for subgraphs
+ * used in the scope of the Graph Network. It translates Subgraphs into Subgraph Versions.
+ * Each version is associated with a Subgraph Deployment. The contract has no knowledge of
+ * human-readable names. All human readable names emitted in events.
+ * The contract implements a multicall behaviour to support batching multiple calls in a single
+ * transaction.
+ * This particular contract is meant to be deployed in L2, and includes helper functions to
+ * receive subgraphs that are migrated from L1.
+ */
+contract L2GNS is GNS, L2GNSV1Storage, IL2GNS {
+    using SafeMathUpgradeable for uint256;
+
+    /// The amount of time (in blocks) that a subgraph owner has to finish the migration
+    /// from L1 before the subgraph can be deprecated: 1 week
+    uint256 public constant FINISH_MIGRATION_TIMEOUT = 50400;
+
+    /// @dev Emitted when a subgraph is received from L1 through the bridge
+    event SubgraphReceivedFromL1(uint256 indexed _subgraphID);
```

Comment:
What do you think about adding the subgraph owner and tokens?

Reply:
Good point, I'll fix it in #764